File: typescript/packages/core/test/thread/rewrite-log.ts
import { expect } from "bun:test";
import { z } from "zod";
import { openStore, storeConnection } from "../../src/agent/sqlite";
import { sha256Hex } from "../../src/hash";
import { canonicalize, type ThreadId } from "../../src/log";
import { unwrap } from "../store/helpers";
import { configHash, type Store } from "./usage-kit";

// Rewrites a stored log the way a writer (or an import) would have produced it: every line
// canonical, re-chained, the head moved, and thread_started's config_hash recomputed when its
// config changed. The result verifies, so a test gets a real log with the shape it needs.

export type Line = Record<string, z.core.util.JSONType>;
const Line = z.record(z.string(), z.json());

/** A JSON object of a log line. */
export function obj(value: z.core.util.JSONType | undefined): Line {
  return Line.parse(value);
}

/** An edit that replaces thread_started's data with `edit(data)`. */
export function editStarted(
  edit: (data: Line) => Line,
): (lines: readonly Line[]) => readonly Line[] {
  return ([started, ...rest]) =>
    started === undefined
      ? []
      : [{ ...started, data: edit(obj(started["data"])) }, ...rest];
}
const Rows = z.array(
  z.object({ seq: z.number(), line: z.instanceof(Uint8Array) }),
);
const Header = z.array(z.object({ header_line: z.instanceof(Uint8Array) }));

/** The pin config_hash covers: thread_started's data but its hash and its parent link. */
function config(line: Line | undefined): Line {
  const { config_hash: _hash, parent: _parent, ...rest } = obj(line?.["data"]);
  return rest;
}

/** Replaces `thread`'s main-branch lines with `edit(lines)` (a prefix of the same length or shorter). */
export async function rewriteLog(
  store: Store,
  thread: ThreadId,
  edit: (lines: readonly Line[]) => readonly Line[],
): Promise<void> {
  const { db } = await storeConnection(store);
  const branch = unwrap((await openStore(store)).log.mainBranch(thread));
  const rows = Rows.parse(
    db.all("SELECT seq, line FROM events WHERE branch_id = ? ORDER BY seq", [
      branch,
    ]),
  );
  const [header] = Header.parse(
    db.all("SELECT header_line FROM branches WHERE branch_id = ?", [branch]),
  );
  const text = new TextDecoder();
  const before = rows.map((r) => Line.parse(JSON.parse(text.decode(r.line))));
  const after = edit(structuredClone(before)).map((line) => ({ ...line }));
  const [started] = after;
  if (
    started !== undefined &&
    configHash(config(started)) !== configHash(config(before[0]))
  ) {
    // Proves the formula on the stored pin before re-hashing the edited one.
    expect(obj(before[0]?.["data"])["config_hash"]).toBe(
      configHash(config(before[0])),
    );
    started["data"] = {
      ...obj(started["data"]),
      config_hash: configHash(config(started)),
    };
  }
  let prev = sha256Hex(header?.header_line ?? new Uint8Array());
  after.forEach((line, i) => {
    const bytes = new TextEncoder().encode(
      unwrap(canonicalize({ ...line, prev_hash: prev })),
    );
    db.run("UPDATE events SET line = ? WHERE branch_id = ? AND seq = ?", [
      bytes,
      branch,
      rows[i]?.seq ?? 0,
    ]);
    prev = sha256Hex(bytes);
  });
  db.run("DELETE FROM events WHERE branch_id = ? AND seq > ?", [
    branch,
    after.length,
  ]);
  db.run(
    "UPDATE branches SET head_seq = ?, head_hash = ? WHERE branch_id = ?",
    [after.length, prev, branch],
  );
}

File: python/tests/thread/rewrite_log.py
"""Rewrites a stored log the way a writer (or an import) would have produced it: every line
canonical, re-chained, the head moved, and thread_started's config_hash recomputed when its config
changed. The result verifies, so a test gets a real log with the shape it needs."""

import copy
import sqlite3
from collections.abc import Callable, Sequence

from pydantic import JsonValue, TypeAdapter

from threads import Store
from threads.agents.store import open_store
from threads.log import ThreadId
from threads.log.digest import sha256_hex
from threads.log.jcs import canonicalize
from threads.result import Ok

type Line = dict[str, JsonValue]
_LINE: TypeAdapter[Line] = TypeAdapter(Line)


def obj(value: JsonValue) -> Line:
    """A JSON object of a log line."""
    return _LINE.validate_python(value)


def config_hash(config: Line) -> str:
    """config_hash of an agent with no extensions, skills, memory, sandbox or subagents: its
    canonical pin."""
    text = canonicalize(config)
    assert isinstance(text, Ok)
    return sha256_hex(text.value.encode())


def _config(line: Line) -> Line:
    """The pin config_hash covers: thread_started's data but its hash and its parent link."""
    return {k: v for k, v in obj(line["data"]).items() if k not in ("config_hash", "parent")}


def edit_started(edit: Callable[[Line], Line]) -> Callable[[list[Line]], list[Line]]:
    """An edit that replaces thread_started's data with `edit(data)`."""

    def apply(lines: list[Line]) -> list[Line]:
        started, *rest = lines
        return [{**started, "data": edit(obj(started["data"]))}, *rest]

    return apply


async def rewrite_log(
    store: Store, thread: ThreadId, edit: Callable[[list[Line]], list[Line]]
) -> None:
    """Replaces `thread`'s main-branch lines with `edit(lines)` (as many lines or fewer)."""
    sq = await open_store(store)
    root = await sq.root(thread)
    assert isinstance(root, Ok)
    branch = root.value

    def rewrite(c: sqlite3.Connection) -> None:
        rows: Sequence[tuple[int, bytes]] = c.execute(
            "SELECT seq, line FROM events WHERE branch_id = ? ORDER BY seq", (branch,)
        ).fetchall()
        (header,) = c.execute(
            "SELECT header_line FROM branches WHERE branch_id = ?", (branch,)
        ).fetchone()
        before = [_LINE.validate_json(line) for _, line in rows]
        after = edit(copy.deepcopy(before))
        if after and _config(after[0]) != _config(before[0]):
            # Proves the formula on the stored pin before re-hashing the edited one.
            assert obj(before[0]["data"])["config_hash"] == config_hash(_config(before[0]))
            data = {**obj(after[0]["data"]), "config_hash": config_hash(_config(after[0]))}
            after[0] = {**after[0], "data": data}
        prev = sha256_hex(header)
        for (seq, _), line in zip(rows, after, strict=False):
            text = canonicalize({**line, "prev_hash": prev})
            assert isinstance(text, Ok)
            raw = text.value.encode()
            c.execute(
                "UPDATE events SET line = ? WHERE branch_id = ? AND seq = ?", (raw, branch, seq)
            )
            prev = sha256_hex(raw)
        c.execute("DELETE FROM events WHERE branch_id = ? AND seq > ?", (branch, len(after)))
        c.execute(
            "UPDATE branches SET head_seq = ?, head_hash = ? WHERE branch_id = ?",
            (len(after), prev, branch),
        )

    await sq.run(rewrite)
